### PR TITLE
Proxy connection and healthcheck

### LINKED ISSUE #108: Add a green/red dot at bottom-right that shows if the proxy is "up"

### LINKED ISSUE #122: After re-opening the tauri app (after succesful pairing) the modal "service unavailable" is thrown
Only way to make it work is to clear cache and pair again.

I think this issue and this #86  can be fixed with #118  along with a refactor of the logic that uses `isReady` RPC as source of truth

### LINKED ISSUE #176: Tauri Dev sidecar not starting and crash
MacBook Pro M1
OS: MacOS Monterey 

Changing `rpc_proxy.go` with `resp.Header().Set("Access-Control-Allow-Origin", "http://localhost:3003")`

When running `cargo tauri dev` the sidecar doesn't start. 
We can start it from `src-tauri/bin/grpcproxy-aarch64-apple-darwin`.
However it quickly crashes.
```
INFO[0000] start listening on :3030                 
INFO[0004] handling gRPC web request: /Operator/ListMarkets 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x240 pc=0x10275b160]

goroutine 23 [running]:
google.golang.org/grpc.(*ClientConn).NewStream(0x0, {0x1029760b8, 0x140000b23c0}, 0x102cca9c0, {0x140000ae095, 0x15}, {0x0, 0x0, 0x0})
	/Users/stephane/go/pkg/mod/google.golang.org/grpc@v1.42.0/stream.go:153 +0x20
google.golang.org/grpc.NewClientStream(...)
	/Users/stephane/go/pkg/mod/google.golang.org/grpc@v1.42.0/stream.go:163
github.com/mwitkow/grpc-proxy/proxy.(*handler).handler(0x1400000e268, {0x0, 0x0}, {0x10297a8e8, 0x140000e00c0})
	/Users/stephane/go/pkg/mod/github.com/mwitkow/grpc-proxy@v0.0.0-20181017164139-0f1106ef9c76/proxy/handler.go:74 +0x120
google.golang.org/grpc.(*Server).processStreamingRPC(0x14000021500, {0x10297e2e0, 0x140000c43f0}, 0x140000e6000, 0x0, 0x140001184c0, 0x0)
	/Users/stephane/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:1540 +0xf44
google.golang.org/grpc.(*Server).handleStream(0x14000021500, {0x10297e2e0, 0x140000c43f0}, 0x140000e6000, 0x0)
	/Users/stephane/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:1626 +0x9a4
google.golang.org/grpc.(*Server).serveStreams.func1.2(0x1400009e210, 0x14000021500, {0x10297e2e0, 0x140000c43f0}, 0x140000e6000)
	/Users/stephane/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:921 +0x94
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/stephane/go/pkg/mod/google.golang.org/grpc@v1.42.0/server.go:919 +0x1f0

Saving session...completed.

[Process completed]
``` 

Running `yarn tauri:build --debug` and changing `rpc_proxy.go` with `resp.Header().Set("Access-Control-Allow-Origin", "tauri://localhost:3003")` works fine.

### LINKED ISSUE #177: [Tauri app] Kill grpcproxy process when quitting
Running app bundled with `cargo tauri build`, quitting app from menu doesn't kill grpcproxy process

### LINKED ISSUE #181: [Proxy] When quitting the app and reopening it, proxy will crash with signal 9 and fails to restart

### PR DESCRIPTION
**Pull request by Janaka-Steph**
Handle proxy connection and health check.
healthCheck is called in useExponentialInterval (fast at startup then every 4 secs).
Fix market selection error by fetching asset data from Esplora and determining market based on ticker rather than asset_id.

It closes #108
It closes #122 
It closes #176 
It closes #177 
It closes #181 

Known issue:
- Quitting app from MacOS Dock will not kill the proxy process because it is not supported by Tauri (https://github.com/tauri-apps/tauri/issues/3084)

Please review @tiero @altafan

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -72,6 +72,7 @@
     "fmt": "yarn run eslint --fix && yarn run prettier --write",
     "lint": "yarn run eslint && yarn run prettier --check",
     "prettier": "prettier './src/**/*.{ts,tsx}'",
+    "proxy:run": "./src-tauri/bin/grpcproxy-aarch64-apple-darwin",
```
Comment (tiero) [inline package.json:75]: how we can generalize for all arch?

Comment (altafan) [reply package.json:75]: `make run` is the answer. It requires golang installed on the machine though.

File: rpc_proxy.go [hunk 2]
```diff
@@ -70,6 +70,11 @@ func (s serviceState) String() string {
 
 func init() {
 	flag.Parse()
+	// Log method name
+	// Adds between between 20 and 40% overhead
```
Comment (altafan) [inline rpc_proxy.go:74]:
```suggestion
	// Adds between 20 and 40% overhead
```

File: src/App.tsx [hunk 3]
```diff
@@ -31,30 +37,97 @@ export const App = (): JSX.Element => {
...
+    console.log('Proxy pid:', child.pid);
+  };
 
-  useEffect(() => {
-    (async () => {
-      if (useProxy && tdexdConnectUrl && proxyIsRunning) {
-        dispatch(connectProxy());
+  const startAndConnectToProxy = useCallback(async () => {
+    console.log('startAndConnectToProxy');
+    if (useProxy && proxyHealth !== 'SERVING') {
```
Comment (altafan) [inline src/App.tsx:54]: better checking the status codes instead of the string description now that they are fixed?

File: src/App.tsx [hunk 4]
```diff
@@ -31,30 +37,97 @@ export const App = (): JSX.Element => {
...
+          }
+        }
+        // If the HTTP call fails
+      } else if (desc === 'NOT_SERVING') {
+        console.error('gRPC Proxy:', desc);
       }
-    })();
-  }, [tdexdConnectUrl, proxyIsRunning, useProxy, dispatch]);
+    }
+  }, [dispatch, macaroonCredentials, proxyHealth, tdexdConnectUrl, useProxy]);
```
Comment (altafan) [inline src/App.tsx:77]: better checking the status codes instead of the string description now that they are fixed?

Comment (Janaka-Steph) [reply src/App.tsx:77]: Since they are not likely to change I preferred using the string to have more explicit code

**Comment by tiero** [comment]
utACK

File: package.json (hunk 1 above)
Comment (Janaka-Steph) [reply package.json]: done

**Comment by altafan** [comment]
@Janaka-Steph I always need to reload the app after pairing with an initialized daemon to see balances, markets, etc correctly. 

From the console, it seems like the dashboard tries to call the grpc methods while connecting. It should either wait for the proxy to be connected or retry all the failing grpc calls.

**Comment by Janaka-Steph** [comment]
> @Janaka-Steph I always need to reload the app after pairing with an initialized daemon to see balances, markets, etc correctly.
> 
> From the console, it seems like the dashboard tries to call the grpc methods while connecting. It should either wait for the proxy to be connected or retry all the failing grpc calls.

Last commit should fix this. Still need to shorten the useInterval loop (fast at startup then 5sec) to speedup the refresh.

**Comment by altafan** [comment]
Yes, it's fixed 👍

File: src/features/home/Home.tsx [hunk 5]
```diff
@@ -12,19 +12,37 @@ import { DashboardPanelRight } from './DashboardPanelRight';
...
   // UnlockWallet Modal
   const [isUnlockWalletModalVisible, setIsUnlockWalletModalVisible] = useState(false);
   const showUnlockWalletModal = () => setIsUnlockWalletModalVisible(true);
   const handleUnlockWalletModalCancel = () => setIsUnlockWalletModalVisible(false);
+  //
+  const [proxyIsServingAndReady, setProxyIsServingAndReady] = useState(false);
+
+  useEffect(() => {
+    console.log('!!!isReady?.isInitialized', isReady?.isInitialized);
+    console.log('!!!proxyHealth', proxyHealth);
```
Comment (tiero) [inline src/features/home/Home.tsx:26]: remove these

Comment (Janaka-Steph) [reply src/features/home/Home.tsx:26]: The PR is not finished.

**Comment by altafan** [comment]
@Janaka-Steph I tested after the last commits:
* the red/yellow/green button signaling the status of the proxy has disappeared if starting the app with `yarn tauri:dev`
* After pairing the dashboard with an initialized daemon, if I quit the tauri app and then restart again with `yarn tauri:dev` the error modal with message `Service Unavailable - Service is not available or credentials are wrong, please retry. ` is shown. It's enough to reload the app to make it work correctly. 
Screenshot 2022-01-10 at 14 55 45

* Dunno if this is a known issue already, but on regtest I'm not able to withdraw funds from a market. I always get the message `market selection error` with any valid amount. Also not sure the select input on the left looks good.
Screenshot 2022-01-10 at 14 39 12

**Comment by Janaka-Steph** [comment]
> Dunno if this is a known issue already, but on regtest I'm not able to withdraw funds from a market. I always get the message market selection error with any valid amount. Also not sure the select input on the left looks good.

Regarding market withdraw error it's because asset ids in state are not correct. You should not have that issue on testnet/mainnet. Anyway, will see how to do better in an other PR.

File: rpc_proxy.go [hunk 6]
```diff
@@ -70,6 +70,11 @@ func (s serviceState) String() string {
 
 func init() {
 	flag.Parse()
+	// Log method name
+	// Adds between 20 and 40% overhead
+	// Comment when not debugging
+	// https://github.com/sirupsen/logrus#logging-method-name
+	// log.SetReportCaller(true)
```
Comment (altafan) [inline rpc_proxy.go:77]: ok, we can get rid of these lines, please delete them.

**Review by altafan — APPROVED** [review]

**Pull request merged**